flink-ml-servable-core/src/main/java/org/apache/flink/ml/util/FileUtils.java

```diff
@@ -0,0 +1,138 @@
+        FileSystem fs = path.getFileSystem();
+        fs.mkdirs(path);
+        return fs;
+    }
+
+    /**
+     * Loads the metadata from the metadata file under the given path.
+     *
+     * <p>The method throws RuntimeException if the expectedClassName is not empty AND it does not
+     * match the className of the previously saved stage.
+     *
+     * @param path The parent directory of the metadata file to read from.
+     * @param expectedClassName The expected class name of the stage.
+     * @return A map from metadata name to metadata value.
+     */
+    public static Map<String, ?> loadMetadata(String path, String expectedClassName)
+            throws IOException {
+        Path metadataPath = new Path(path, "metadata");
+        FileSystem fs = metadataPath.getFileSystem();
+
+        StringBuilder buffer = new StringBuilder();
+        try (BufferedReader br = new BufferedReader(new InputStreamReader(fs.open(metadataPath)))) {
+            String line;
+            while ((line = br.readLine()) != null) {
+                if (!line.startsWith("#")) {
+                    buffer.append(line);
+                }
+            }
+        }
+
+        @SuppressWarnings("unchecked")
+        Map<String, ?> result = JsonUtils.OBJECT_MAPPER.readValue(buffer.toString(), Map.class);
+
+        String className = (String) result.get("className");
+        if (!expectedClassName.isEmpty() && !expectedClassName.equals(className)) {
+            throw new RuntimeException(
+                    "Class name "
+                            + className
+                            + " does not match the expected class name "
+                            + expectedClassName
+                            + ".");
+        }
+
+        return result;
+    }
+
+    // Returns a string with value {parentPath}/stages/{stageIdx}, where the stageIdx is prefixed
+    // with zero or more `0` to have the same length as numStages. The resulting string can be
+    // used as the directory to save a stage of the Pipeline or PipelineModel.
+    public static String getPathForPipelineStage(int stageIdx, int numStages, String parentPath) {
+        String format =
+                String.format("stages%s%%0%dd", File.separator, String.valueOf(numStages).length());
+        String fileName = String.format(format, stageIdx);
+        return new Path(parentPath, fileName).toString();
+    }
+
+    /** Returns a subdirectory of the given path for saving/loading model data. */
+    public static Path getDataPath(String path) {
+        return new Path(path, "data");
+    }
+
+    /**
+     * Opens an FSDataInputStream to read the model data file in the directory. Only one model data
+     * file is expected to be in the directory.
+     *
+     * @param path The parent directory of the model data file.
+     * @return A FSDataInputStream to read the model data.
+     */
+    public static FSDataInputStream getModelDataInputStream(Path path) throws IOException {
```

Comment:
Yes, you are right. Besides, I moved this function to `ServableReadWriteUtils.java` because I expect that all util methods that are required for implementing an algorithm are located in `ServableReadWriteUtils.java`, the `FileUtils.java` is only for the framework.